plugins/upower.c

```diff
@@ -0,0 +1,471 @@
+
+	DBG("calling set_indicator: %d", val);
+	ofono_emulator_set_indicator(em, OFONO_EMULATOR_IND_BATTERY, (int) val);
+}
+
+static gboolean update_battery_level(double percentage_val)
+{
+	GList *list;
+	int battery_level;
+
+	if (percentage_val <= 1.00) {
+		battery_level = 0;
+	} else if (percentage_val > 1.00 && percentage_val <= 100.00) {
+		battery_level = ((int) percentage_val - 1) / 20 + 1;
+	} else {
+		ofono_error("%s: Invalid value for battery level: %f",
+								__func__,
+								percentage_val);
+		return FALSE;
+	}
+
+	DBG("last_battery_level: %d battery_level: %d", last_battery_level,
+							battery_level);
+
+	if (last_battery_level == battery_level)
+		goto done;
+	else
+		last_battery_level = battery_level;
+
+	for (list = modems; list; list = list->next) {
+		struct ofono_modem *modem = list->data;
+
+		__ofono_modem_foreach_registered_atom(modem,
+					OFONO_ATOM_TYPE_EMULATOR_HFP,
+					emulator_battery_cb,
+					GINT_TO_POINTER(battery_level));
+	}
+done:
+	return TRUE;
+}
+
+static gboolean battery_props_changed(DBusConnection *conn, DBusMessage *msg,
+				void *user_data)
+
+{
+	const char *iface;
+	DBusMessageIter iter, dict;
+	double percentage_val;
+	gboolean percentage_found = FALSE;
+
+	DBG("");
+
+	dbus_message_iter_init(msg, &iter);
+
+	if (dbus_message_iter_get_arg_type(&iter) != DBUS_TYPE_STRING) {
+		ofono_error("%s: iface != TYPE_STRING!", __func__);
+		return FALSE;
+	}
+
+	dbus_message_iter_get_basic(&iter, &iface);
+
+	if (g_str_equal(iface, UPOWER_DEVICE_INTERFACE) != TRUE) {
+		ofono_error("%s: wrong iface: %s!", __func__, iface);
+		return FALSE;
+	}
+
+	if (!dbus_message_iter_next(&iter)) {
+		ofono_error("%s: advance iter failed!", __func__);
+		return FALSE;
+	}
+
+	if (dbus_message_iter_get_arg_type(&iter) != DBUS_TYPE_ARRAY) {
+		ofono_error("%s: type != ARRAY!", __func__);
+		return FALSE;
+	}
+
+	dbus_message_iter_recurse(&iter, &dict);
+
+	while (dbus_message_iter_get_arg_type(&dict) == DBUS_TYPE_DICT_ENTRY) {
+		DBusMessageIter entry, val;
+		const char *key;
+
+		dbus_message_iter_recurse(&dict, &entry);
+
+		if (dbus_message_iter_get_arg_type(&entry) !=
+				DBUS_TYPE_STRING) {
+			ofono_error("%s: key type != STRING!", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_get_basic(&entry, &key);
+
+		if (g_str_equal(key, "Percentage") != TRUE) {
+			dbus_message_iter_next(&dict);
+			continue;
+		}
+
+		dbus_message_iter_next(&entry);
+		if (dbus_message_iter_get_arg_type(&entry) !=
+				DBUS_TYPE_VARIANT) {
+			ofono_error("%s: 'Percentage' val != VARIANT",
+								__func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_recurse(&entry, &val);
+
+		if (dbus_message_iter_get_arg_type(&val) != DBUS_TYPE_DOUBLE) {
+			ofono_error("%s: 'Percentage' val != DOUBLE", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_get_basic(&val, &percentage_val);
+		percentage_found = TRUE;
+		break;
+	}
+
+	/* Other properties can change, so don't trigger cb removal */
```

Comment:
Maybe we should always return TRUE beyond this point? Removing the signal watch because battery level was, say, 100.01% seems a bit picky. Of course should not happen, but...